### PR TITLE
Help Center: Add tailored article list

### LINKED ISSUE #88707: Help Center: Extend contextual help with tailored posts
Context: https://github.com/Automattic/wp-calypso/issues/88555#issuecomment-2008597765

In the Help Center we offer contextual docs based on the url we are in. We have this list here: https://github.com/Automattic/wp-calypso/blob/trunk/packages/help-center/src/route-to-query-mapping.json

It would be nice also to offer the possibility of specifying a tailored list of post ids to show when on a certain url.

Check if tailored post ids exist should happen here: https://github.com/Automattic/wp-calypso/blob/trunk/packages/help-center/src/hooks/use-context-based-search-mapping.tsx#L13

Take also locale into consideration. (different support sites have different ids for the same article in different languages)

Algo could be:
1. Search for a tailored list of ids, if it exists show it.
2. Otherwise fallback to contextual search as before.

### PR DESCRIPTION
**Pull request by renancarvalho**
Fixes https://github.com/Automattic/wp-calypso/issues/88707

## Proposed Changes

* This https://github.com/Automattic/wp-calypso/pull/88808 added a feature that allows tailored articles suggestions. This current PR is adding the article's structure to be suggested.
c/c @donalirl 

## Testing Instructions

* Using live link 
* Navigate to the pages (https://github.com/Automattic/wp-calypso/pull/89471/files#diff-1387095f7173d264c294187f77c20df1f285bad3ad7262e615718598da23ed65R1) 
* Open help center
* Make sure the articles suggested are matching the https://github.com/Automattic/wp-calypso/pull/89471/files#diff-1387095f7173d264c294187f77c20df1f285bad3ad7262e615718598da23ed65R1

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Review by escapemanuele — CHANGES_REQUESTED** [review]
Hey @renancarvalho, I get an error in pages with fewer than 5 posts (like `/import?`

```83029.a0b0c1d….min.js:25 TypeError: tD[a]?.find is not a function```

**Comment by renancarvalho** [comment]
> Hey @renancarvalho, I get an error in pages with fewer than 5 posts (like `/import?`
> 
> `83029.a0b0c1d….min.js:25 TypeError: tD[a]?.find is not a function`

Great catch! It was the structure using {} instead of []. Pushed the fix <3

**Review by escapemanuele — CHANGES_REQUESTED** [review]
Found this!

image

On `/import` we have duplicates. So it seems we're not removing duplicates when we have both tailored articles and search query :) We should do that.

**Comment by renancarvalho** [comment]
> Found this!
> 
> image
> On `/import` we have duplicates. So it seems we're not removing duplicates when we have both tailored articles and search query :) We should do that.

Great! I have added a `filterOut` function to remove duplicated articles

File: client/data/help/use-help-search-query.ts
Comment (escapemanuele) [inline client/data/help/use-help-search-query.ts:80]: Pay attention that this is confusing. You pass articles and then search, while the function parameters are search and then articles.
I spent a little trying to understand why it worked, before noticing this swap 😄

**Comment by escapemanuele** [comment]
Left a comment regarding parameters naming to be changed :)

One thing that we can think of, in the future... Would it be possible to have only the locale or the blog id? :)

**Comment by renancarvalho** [comment]
> Left a comment regarding parameters naming to be changed :)
> 
> One thing that we can think of, in the future... Would it be possible to have only the locale or the blog id? :)

I think we could think of removing locale since support sites have different blog_id's per “locale”, need more profound investigation, though. Not sure about dev sites.

**Comment by escapemanuele** [comment]
Sure thing, let's investigate when we have some down time! 

Opened this issue to track this: https://github.com/Automattic/wp-calypso/issues/89524

**Review by escapemanuele — APPROVED** [review]
Thanks for addressing the feedback! Let's wait for @donalirl and then merge this 👍

**Comment by donalirl** [comment]
Overall it's working really well! I checked every screen and the correct pages loaded, except for these ones:

- me/purchases
- me/purchases/billing
- me/purchases/payment-methods
- me/purchases/subscriptions/ (possibly, I don't have this tab)
- marketing/tools

They loaded blank:
Screenshot 2024-04-16 at 10 36 13

Two other things:
* I wasn't able to check the WP-Admin screens, as they don't seem to work within the Calypso testing environment
* Is it possible to set articles for the editor screens? Those are:
  * /page/
  * /post/
  * /wp-admin/site-editor.php

**Comment by renancarvalho** [comment]
> * , as they don't seem to work within the Calypso testing environm

Hi @donalirl, thank you for testing this.

I noticed that some IDs don't have a post related to it or to the blog, in this case, the API was failing, I have fixed it, now if the post_Id does not exist it is just ignored. 👍 

Regarding the WP-admin pages, it is possible by updating a testing version of the ETK plugin on your atomic site, I can send you instructions on how to do that.

Finally, yes, it should be possible to set articles for `/page/, /post/, /wp-admin/site-editor.php` , should be the same as the other pages.

🙇

**Comment by renancarvalho** [comment]
@donalirl I have tested on my own Atomic site the `wp-admin` URL's

On `wp-admin/edit.php?post_type=page` this is what I get and they match the ids you have added

image


On `/wp-admin/edit.php` also matching the ids you have added

image


On `wp-admin/post-new.php?post_type=page` also matching the ids you have added

image

On `wp-admin/post-new.php` also matching the ids you have added

image

**Comment by escapemanuele** [comment]
@donalirl an example of articles not working for me are 276100 and 111341

**Comment by donalirl** [comment]
> I noticed that some IDs don't have a post related to it or to the blog, in this case, the API was failing, I have fixed it, now if the post_Id does not exist it is just ignored. 👍

Do you have a record of which IDs those were? It's possible I typed it wrong. For the ones @escapemanuele shared above, and the ones mentioned in Slack:

276100 - this should be 267100
111341 - this should be 111349
3077 - this should be 3307

> Finally, yes, it should be possible to set articles for /page/, /post/, /wp-admin/site-editor.php , should be the same as the other pages.

Great, can we add:

```
	"/page/": [ { "locale": "en", "post_ids": [ 147594, 193416, 193554, 167011, 193903 ], "blog_id": 9619154 } ],
	"/post/": [ { "locale": "en", "post_ids": [ 147594, 193416, 193554, 167011, 176675 ], "blog_id": 9619154 } ],
	"/wp-admin/site-editor.php/": [ { "locale": "en", "post_ids": [ 188667, 193554, 162159, 192200, 176675 ], "blog_id": 9619154 } ],

```

**Comment by renancarvalho** [comment]
@donalirl, thank you for providing the new IDs

I have updated the new IDs and added the new pages 👍 
 
The only weird one is `wp-admin/site-editor.php` that took me to this page with the old Help Center icon (FAB if I am not mistaken)

image

@escapemanuele I think we should take a look at this URL in another task, WDYT? If that is fine, I think we are good to merge this one.

**Comment by escapemanuele** [comment]
I think we can deploy this, thanks a lot 👍

**Pull request merged**
